agent/command/s3_copy.go

```diff
@@ -188,43 +243,70 @@ func (c *s3copy) s3Copy(ctx context.Context,
-			AwsSecret:           c.AwsSecret,
-			S3SourceRegion:      s3CopyFile.Source.Region,
-			S3SourceBucket:      s3CopyFile.Source.Bucket,
-			S3SourcePath:        s3CopyFile.Source.Path,
-			S3DestinationRegion: s3CopyFile.Destination.Region,
-			S3DestinationBucket: s3CopyFile.Destination.Bucket,
-			S3DestinationPath:   s3CopyFile.Destination.Path,
-			S3DisplayName:       s3CopyFile.DisplayName,
-			S3Permissions:       s3CopyFile.Permissions,
-		}
-
-		responseString, err := comm.S3Copy(ctx, td, &s3CopyReq)
+		s3CopyReq.AwsKey = s3pc.AwsKey
+		s3CopyReq.AwsSecret = s3pc.AwsSecret
 
-		if responseString != "" {
-			logger.Task().Infof("s3Copy response: %s", responseString)
+		// Now copy the file into the permanent location
+		srcOpts := pail.S3Options{
+			Credentials: pail.CreateAWSCredentials(s3CopyReq.AwsKey, s3CopyReq.AwsSecret, ""),
+			Region:      s3CopyReq.S3SourceRegion,
+			Name:        s3CopyReq.S3SourceBucket,
+			Permissions: pail.S3Permissions(s3CopyReq.S3Permissions),
 		}
 
+		srcBucket, err := pail.NewS3MultiPartBucketWithHTTPClient(client, srcOpts)
 		if err != nil {
-			err = errors.Wrap(err, "s3 push copy failed")
-			logger.Execution().Error(err)
-
-			if s3CopyFile.Optional {
-				logger.Execution().Errorf("file '%s' is optional, continuing",
-					s3CopyFile.DisplayName)
-				continue
-			} else {
-				return errors.Wrapf(err, "failed to push %s to %s",
-					s3CopyFile.Source.Path, s3CopyFile.Destination.Bucket)
-			}
-
+			grip.Error(errors.Wrap(err, "S3 copy failed, could not establish connection to source bucket"))
 		}
 
-		err = c.attachFiles(ctx, comm, logger, td, s3CopyReq)
+		if err := srcBucket.Check(ctx); err != nil {
+			return errors.Wrap(err, "invalid bucket")
+		}
+		destOpts := pail.S3Options{
+			Credentials: pail.CreateAWSCredentials(s3CopyReq.AwsKey, s3CopyReq.AwsSecret, ""),
+			Region:      s3CopyReq.S3DestinationRegion,
+			Name:        s3CopyReq.S3DestinationBucket,
+			Permissions: pail.S3Permissions(s3CopyReq.S3Permissions),
+		}
+		destBucket, err := pail.NewS3MultiPartBucket(destOpts)
 		if err != nil {
-			return errors.WithStack(err)
+			grip.Error(errors.Wrap(err, "S3 copy failed, could not establish connection to destination bucket"))
+		}
+
+		for i := 1; i <= maxS3OpAttempts; i++ {
+			select {
+			case <-ctx.Done():
+				return errors.New("s3 put operation canceled")
+			case <-timer.C:
+				copyOpts := pail.CopyOptions{
+					SourceKey:         s3CopyReq.S3SourcePath,
+					DestinationKey:    s3CopyReq.S3DestinationPath,
+					DestinationBucket: destBucket,
+				}
+				err = srcBucket.Copy(ctx, copyOpts)
+				if err != nil {
+					newPushLog.Status = pushLogFailed
+					if err := comm.UpdatePushStatus(ctx, td, newPushLog); err != nil {
+						return errors.Wrap(err, "updating pushlog status failed for task")
+					}
+					if s3CopyFile.Optional {
+						logger.Execution().Errorf("S3 push copy failed to copy'%s' to '%s'. File is optional, continuing \n error",
```

Comment:
nit: missing space between copy and '%s'. Also not sure what this `\n error` is for

Reply:
\n error makes it more readable in the task log. It tends to get pretty long, this puts err: on a new line so the user doesn't have to keep scrolling horizontally.

Reply:
Oh did you mean to log the actual error? This is just logging the word error.

Reply:
Ah, I meant to wrap the error. I'll add it, thanks!